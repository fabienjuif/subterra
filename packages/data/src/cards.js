export const EndCard = {
  type: 'end',
  id: 'hKyJQmhBBSgFdJkZStisD',
}
export const WaterCard = {
  type: 'water',
  damage: 2,
  duration: 1,
  name: 'water',
  id: 'jVjrsEynaxHyQ0o2bn3F5',
}
export const WaterCard2 = {
  type: 'water',
  damage: 2,
  duration: 2,
  name: 'water 2',
  id: 'mQsb0CISJMDosYAl1y9bK',
}
export const GazCard = {
  type: 'gaz',
  damage: 2,
  duration: 1,
  name: 'gaz',
  id: '3KNfiVS61yBJ2lE26OR0N',
}
export const GazCard2 = {
  type: 'gaz',
  damage: 2,
  duration: 2,
  name: 'gaz 2',
  id: '0vK8Hw4hKFPRGf76BB7Nl',
}
export const LandslideCard = {
  type: 'landslide',
  damage: 3,
  duration: 1,
  name: 'landslide',
  id: 'XQeHmlrBzuZTpt2SXvVZe',
}
export const LandslideCard2 = {
  type: 'landslide',
  damage: 3,
  duration: 2,
  name: 'landslide 2',
  id: 'MxWpXeRkxDUK20CdwGpqD',
}
export const ShakeCard = {
  type: 'shake',
  damage: 1,
  duration: 1,
  name: 'shake',
  id: '1sKrNtKjjrYkrSDdmWkYy',
}
export const HorrorCard = {
  type: 'enemy',
  damage: 3,
  duration: 1,
  name: 'horror',
  id: 'Xk0FDZpwRm2re3Uox5Qg4',
}
export const HorrorCard2 = {
  type: 'enemy',
  damage: 3,
  duration: 2,
  name: 'horror 2',
  id: 'RVdgxn0bohDJuePSe8Na7',
}

const cards = [
  { card: GazCard, remaining: 4 },
  { card: HorrorCard, remaining: 4 },
  { card: LandslideCard, remaining: 5 },
  { card: ShakeCard, remaining: 7 },
  { card: WaterCard, remaining: 5 },
  { card: GazCard2, remaining: 1 },
  { card: HorrorCard2, remaining: 1 },
  { card: LandslideCard2, remaining: 1 },
  { card: WaterCard2, remaining: 1 },
]

export default cards
