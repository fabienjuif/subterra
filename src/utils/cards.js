const EndCard = { type: 'end' }
const WaterCard = { type: 'water', damage: 2, duration: 1 }
const WaterCard2 = { type: 'water', damage: 2, duration: 2 }
const GazCard = { type: 'gaz', damage: 2, duration: 1 }
const GazCard2 = { type: 'gaz', damage: 2, duration: 2 }
const LandslideCard = { type: 'landslide', damage: 3, duration: 1 }
const LandslideCard2 = { type: 'landslide', damage: 3, duration: 2 }
const ShakeCard = { type: 'shake', damage: 1, duration: 1 }
const HorrorCard = { type: 'enemy', damage: 0, duration: 1 }
const HorrorCard2 = { type: 'enemy', damage: 0, duration: 2 }

const cards = [
  EndCard,
  GazCard,
  LandslideCard2,
  ShakeCard,
  ShakeCard,
  GazCard,
  WaterCard,
  HorrorCard,
  LandslideCard,
  WaterCard,
  HorrorCard,
  WaterCard2,
  LandslideCard,
  WaterCard,
  ShakeCard,
  HorrorCard2,
  ShakeCard,
  ShakeCard,
  ShakeCard,
  GazCard,
  GazCard2,
  LandslideCard,
  WaterCard,
  LandslideCard,
  ShakeCard,
  LandslideCard,
  HorrorCard,
  WaterCard,
  GazCard,
  HorrorCard,
]

export default cards
