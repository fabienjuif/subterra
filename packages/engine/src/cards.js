import { EndCard } from '@subterra/data'
import { players, roll } from './actions'
import { tiles, random } from './utils'

export const init = (store, action) => {
  store.mutate((state) => {
    state.cards = action.payload
  })
}

export const pick = (store, action) => {
  store.mutate((state) => {
    // draw a new card
    // - if there is 1 remaining card we draw a "end" card
    // - in other case take a card from remaining one
    //    and remove cards from deck when there is no more remaining
    let nextCard
    state.cards.remaining = Math.max(0, state.cards.remaining - 1)
    if (state.cards.remaining <= 0) {
      nextCard = { ...EndCard }
    } else {
      const { value, nextSeed } = random.roll(
        state.cards.deck.length,
        state.seeds.cardsNext,
      )
      state.seeds.cardsNext = nextSeed

      const cardInDeck = state.cards.deck[value - 1]
      cardInDeck.remaining -= 1
      if (cardInDeck.remaining <= 0) {
        state.cards.deck.splice(value - 1, 1)
      }

      nextCard = { ...cardInDeck.card }
    }
    state.cards.active = nextCard
  })

  const nextState = store.getState()
  const { type: cardType } = nextState.cards.active
  if (['shake', 'water', 'gaz', 'enemy', 'end'].includes(cardType)) {
    store.dispatch({
      type: `@cards>${cardType}`,
      payload: { card: nextState.cards.active },
    })
  } else if (cardType === 'landslide') {
    store.dispatch(roll.then({ type: '@cards>landslide' }))
  }
}

export const end = (store, action) => {
  store.getState().players.forEach((player) => {
    if (player.health <= 0) return

    store.dispatch(
      roll.failThen(
        3,
        player,
        players.damage(player, 1000, {
          from: { card: action.payload.card },
        }),
      ),
    )
  })
}

export const shake = (store, action) => {
  const previousState = store.getState()

  previousState.players.forEach((player) => {
    store.dispatch(
      roll.failThen(
        4,
        player,
        players.damage(player, previousState.cards.active.damage, {
          card: previousState.cards.active,
        }),
      ),
    )
  })
}

export const landslide = (store, action) => {
  const {
    cards: { active },
  } = store.getState()

  // find all tiles that are landslide and match the dice result
  // tile should not be already in the landslide status
  // add the status 'landslide' to these tiles
  // and for each of these tiles, check a player is in it and damage it in this case
  store.mutate((state) => {
    state.grid.forEach((tile) => {
      const { type, status, dices } = tile

      if (type !== 'landslide') return
      if (status.includes('landslide')) return
      if (!dices.includes(action.payload.rolled)) return

      tile.status.push('landslide')

      state.players.forEach((player) => {
        if (!tiles.isCellEqual(player)(tile)) return

        store.dispatch(
          players.damage(player, active.damage, {
            card: active,
          }),
        )
      })
    })
  })
}

export const processMarkerCard = (store, action) => {
  const { card } = action.payload

  // find all tiles that have water type and put a status on it
  // if it do not already exists
  // if a player is in this tile then it take damage
  store.mutate((state) => {
    state.grid.forEach((tile) => {
      const { type, status } = tile

      if (type !== card.type) return
      if (status.includes(card.type)) return

      tile.status.push(card.type)

      state.players.forEach((player) => {
        if (!tiles.isCellEqual(player)(tile)) return

        store.dispatch(
          players.damage(player, card.damage, {
            card,
          }),
        )
      })
    })
  })
}
