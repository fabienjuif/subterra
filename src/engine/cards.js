import { players, roll } from './actions'
import { isCellEqual } from '../utils/tiles'

export const init = (store, action) => {
  store.mutate((state) => {
    state.deckCards = action.payload
  })
}

export const pick = (store, action) => {
  store.mutate((state) => {
    if (state.deckCards.length > 0) {
      state.activeCard = state.deckCards.shift()
    }
  })

  const nextState = store.getState()
  const { type: cardType } = nextState.activeCard
  if (['shake', 'water', 'gaz', 'enemy'].includes(cardType)) {
    store.dispatch(`@cards>${cardType}`)
  } else if (nextState.activeCard.type === 'landslide') {
    store.dispatch(roll.then({ type: '@cards>landslide' }))
  }
}

export const shake = (store, action) => {
  const previousState = store.getState()

  previousState.players.forEach((player) => {
    store.dispatch(
      roll.failThen(
        4,
        player,
        players.damage(player, previousState.activeCard.damage, {
          card: previousState.activeCard,
        }),
      ),
    )
  })
}

export const landslide = (store, action) => {
  const { activeCard } = store.getState()

  // find all tiles that are landslide and match the dice result
  // tile should not be already in the landslide status
  // add the status 'landslide' to these tiles
  // and for each of these tiles, check a player is in it and damage it in this case
  store.mutate((state) => {
    state.grid.forEach((tile) => {
      const { type, status, dices } = tile

      if (type !== 'landslide') return
      if (status.includes('landslide')) return
      if (!dices.includes(action.payload.rolled)) return

      tile.status.push('landslide')

      state.players.forEach((player) => {
        if (!isCellEqual(player)(tile)) return

        store.dispatch(
          players.damage(player, activeCard.damage, {
            card: activeCard,
          }),
        )
      })
    })
  })
}

export const processMarkerCard = (store, action) => {
  const { card } = action.payload

  // find all tiles that have water type and put a status on it
  // if it do not already exists
  // if a player is in this tile then it take damage
  store.mutate((state) => {
    state.grid.forEach((tile) => {
      const { type, status } = tile

      if (type !== card.type) return
      if (status.includes(card.type)) return

      tile.status.push(card.type)

      state.players.forEach((player) => {
        if (!isCellEqual(player)(tile)) return

        store.dispatch(
          players.damage(player, card.damage, {
            card,
          }),
        )
      })
    })
  })
}
