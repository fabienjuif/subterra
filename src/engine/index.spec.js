import { createPatch } from 'diff'
import { createEngine } from './index'
import cards, {
  EndCard,
  GazCard,
  WaterCard,
  ShakeCard,
  HorrorCard,
} from '../utils/cards'

expect.addSnapshotSerializer({
  test: () => true,
  print: (str) => str,
})

describe('engine without mock', () => {
  it('should be replayable', () => {
    const play = () => {
      const engine = createEngine()

      engine.dispatch({
        type: '@cards>init',
        payload: [cards[2], cards[3], cards[1], cards[0]],
      })
      engine.dispatch('@players>init')
      engine.dispatch('@cards>pick')
      // TODO: here play a fake game

      return engine
    }

    // 3 times should be enough
    const firstState = play().getState()
    expect(play().getState()).toEqual(firstState)
    expect(play().getState()).toEqual(firstState)
  })

  it('should connect behaviours to actions', () => {
    const engine = createEngine()

    const dispatchAndSnap = (action) => {
      const previousState = engine.getState()
      engine.dispatch(action)
      const afterState = engine.getState()

      expect(
        createPatch(
          typeof action === 'object' ? action.type : action,
          JSON.stringify(previousState, null, 2),
          JSON.stringify(afterState, null, 2),
        ),
      ).toMatchSnapshot()
    }

    dispatchAndSnap({
      type: '@cards>init',
      payload: [ShakeCard, GazCard, WaterCard, HorrorCard, EndCard],
    })
    dispatchAndSnap('@players>init')
    dispatchAndSnap({ type: '@dices>init', payload: [2, 3, 4, 5] })
    dispatchAndSnap('@cards>pick')

    // TODO: complete this after we can discover and move (or explore)
  })
})
