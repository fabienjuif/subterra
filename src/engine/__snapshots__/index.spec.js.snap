// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`engine without mock should connect behaviours to actions 1`] = `
Index: @cards>init
===================================================================
--- @cards>init
+++ @cards>init
@@ -4,9 +4,33 @@
   "enemies": [],
   "deckTiles": {
     "length": 10
   },
-  "deckCards": [],
+  "deckCards": [
+    {
+      "type": "shake",
+      "damage": 1,
+      "duration": 1
+    },
+    {
+      "type": "gaz",
+      "damage": 2,
+      "duration": 1
+    },
+    {
+      "type": "water",
+      "damage": 2,
+      "duration": 1
+    },
+    {
+      "type": "enemy",
+      "damage": 3,
+      "duration": 1
+    },
+    {
+      "type": "end"
+    }
+  ],
   "dices": [],
   "activeCard": {},
   "grid": [
     {
@@ -25,7 +49,36 @@
     "current": {},
     "possibilities": []
   },
   "technical": {
-    "actions": []
+    "actions": [
+      {
+        "type": "@cards>init",
+        "payload": [
+          {
+            "type": "shake",
+            "damage": 1,
+            "duration": 1
+          },
+          {
+            "type": "gaz",
+            "damage": 2,
+            "duration": 1
+          },
+          {
+            "type": "water",
+            "damage": 2,
+            "duration": 1
+          },
+          {
+            "type": "enemy",
+            "damage": 3,
+            "duration": 1
+          },
+          {
+            "type": "end"
+          }
+        ]
+      }
+    ]
   }
 }
\\ No newline at end of file

`;

exports[`engine without mock should connect behaviours to actions 2`] = `
Index: @players>init
===================================================================
--- @players>init
+++ @players>init
@@ -1,7 +1,37 @@
 {
   "gameOver": false,
-  "players": [],
+  "players": [
+    {
+      "id": 0,
+      "x": 0,
+      "y": 0,
+      "health": 3,
+      "name": "Sutat",
+      "archetype": "explorer",
+      "actionPoints": 2,
+      "current": true,
+      "first": true
+    },
+    {
+      "id": 1,
+      "x": 0,
+      "y": 0,
+      "health": 3,
+      "name": "Tripa",
+      "archetype": "chef",
+      "actionPoints": 2
+    },
+    {
+      "id": 2,
+      "x": 0,
+      "y": 0,
+      "health": 5,
+      "name": "SoE",
+      "archetype": "miner",
+      "actionPoints": 2
+    }
+  ],
   "enemies": [],
   "deckTiles": {
     "length": 10
   },
@@ -77,8 +107,11 @@
           {
             "type": "end"
           }
         ]
+      },
+      {
+        "type": "@players>init"
       }
     ]
   }
 }
\\ No newline at end of file

`;

exports[`engine without mock should connect behaviours to actions 3`] = `
Index: @dices>init
===================================================================
--- @dices>init
+++ @dices>init
@@ -59,9 +59,14 @@
     {
       "type": "end"
     }
   ],
-  "dices": [],
+  "dices": [
+    2,
+    3,
+    4,
+    5
+  ],
   "activeCard": {},
   "grid": [
     {
       "id": 0,
@@ -110,8 +115,17 @@
         ]
       },
       {
         "type": "@players>init"
+      },
+      {
+        "type": "@dices>init",
+        "payload": [
+          2,
+          3,
+          4,
+          5
+        ]
       }
     ]
   }
 }
\\ No newline at end of file

`;

exports[`engine without mock should connect behaviours to actions 4`] = `
Index: @cards>pick
===================================================================
--- @cards>pick
+++ @cards>pick
@@ -15,18 +15,18 @@
     {
       "id": 1,
       "x": 0,
       "y": 0,
-      "health": 3,
+      "health": 2,
       "name": "Tripa",
       "archetype": "chef",
       "actionPoints": 2
     },
     {
       "id": 2,
       "x": 0,
       "y": 0,
-      "health": 5,
+      "health": 4,
       "name": "SoE",
       "archetype": "miner",
       "actionPoints": 2
     }
@@ -36,13 +36,8 @@
     "length": 10
   },
   "deckCards": [
     {
-      "type": "shake",
-      "damage": 1,
-      "duration": 1
-    },
-    {
       "type": "gaz",
       "damage": 2,
       "duration": 1
     },
@@ -60,14 +55,15 @@
       "type": "end"
     }
   ],
   "dices": [
-    2,
-    3,
-    4,
     5
   ],
-  "activeCard": {},
+  "activeCard": {
+    "type": "shake",
+    "damage": 1,
+    "duration": 1
+  },
   "grid": [
     {
       "id": 0,
       "type": "start",
@@ -124,8 +120,171 @@
           3,
           4,
           5
         ]
+      },
+      {
+        "type": "@cards>pick"
+      },
+      {
+        "type": "@cards>shake"
+      },
+      {
+        "type": "@dices>roll",
+        "payload": {
+          "min": 4,
+          "playerName": "SoE",
+          "actionOnFail": {
+            "type": "@players>damage",
+            "payload": {
+              "damage": 1,
+              "from": {
+                "card": {
+                  "type": "shake",
+                  "damage": 1,
+                  "duration": 1
+                }
+              },
+              "playerName": "SoE"
+            }
+          }
+        }
+      },
+      {
+        "type": "@dices>rolled",
+        "payload": {
+          "min": 4,
+          "playerName": "SoE",
+          "actionOnFail": {
+            "type": "@players>damage",
+            "payload": {
+              "damage": 1,
+              "from": {
+                "card": {
+                  "type": "shake",
+                  "damage": 1,
+                  "duration": 1
+                }
+              },
+              "playerName": "SoE"
+            }
+          },
+          "value": 2
+        }
+      },
+      {
+        "type": "@players>damage",
+        "payload": {
+          "damage": 1,
+          "from": {
+            "card": {
+              "type": "shake",
+              "damage": 1,
+              "duration": 1
+            }
+          },
+          "playerName": "SoE"
+        }
+      },
+      {
+        "type": "@dices>roll",
+        "payload": {
+          "min": 4,
+          "playerName": "Tripa",
+          "actionOnFail": {
+            "type": "@players>damage",
+            "payload": {
+              "damage": 1,
+              "from": {
+                "card": {
+                  "type": "shake",
+                  "damage": 1,
+                  "duration": 1
+                }
+              },
+              "playerName": "Tripa"
+            }
+          }
+        }
+      },
+      {
+        "type": "@dices>rolled",
+        "payload": {
+          "min": 4,
+          "playerName": "Tripa",
+          "actionOnFail": {
+            "type": "@players>damage",
+            "payload": {
+              "damage": 1,
+              "from": {
+                "card": {
+                  "type": "shake",
+                  "damage": 1,
+                  "duration": 1
+                }
+              },
+              "playerName": "Tripa"
+            }
+          },
+          "value": 3
+        }
+      },
+      {
+        "type": "@players>damage",
+        "payload": {
+          "damage": 1,
+          "from": {
+            "card": {
+              "type": "shake",
+              "damage": 1,
+              "duration": 1
+            }
+          },
+          "playerName": "Tripa"
+        }
+      },
+      {
+        "type": "@dices>roll",
+        "payload": {
+          "min": 4,
+          "playerName": "Sutat",
+          "actionOnFail": {
+            "type": "@players>damage",
+            "payload": {
+              "damage": 1,
+              "from": {
+                "card": {
+                  "type": "shake",
+                  "damage": 1,
+                  "duration": 1
+                }
+              },
+              "playerName": "Sutat"
+            }
+          }
+        }
+      },
+      {
+        "type": "@dices>rolled",
+        "payload": {
+          "min": 4,
+          "playerName": "Sutat",
+          "actionOnFail": {
+            "type": "@players>damage",
+            "payload": {
+              "damage": 1,
+              "from": {
+                "card": {
+                  "type": "shake",
+                  "damage": 1,
+                  "duration": 1
+                }
+              },
+              "playerName": "Sutat"
+            }
+          },
+          "value": 4
+        }
       }
     ]
   }
 }
\\ No newline at end of file

`;
